### PR TITLE
Migrated HLSL 2016 with SM 6_0

### PR DESCRIPTION
**Pull request by toddmacedo**

### CODE REVIEW

File: desktop-src/direct3dhlsl_current/hlsl-landing-page.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# High-level shader language (HLSL)
```
Comment (tex3d) [inline desktop-src/direct3dhlsl_current/hlsl-landing-page.md:1]: Let's describe and link the sub-headings under this landing page in the TOC.  Let's do the same for the section pages (adding pages if necessary).  We can follow the example of the legacy HLSL landing/section pages.

The goal is to help people find the documentation they are looking for by summarizing the sections and the types of content found within them.  Point out when certain types of content divisions might not be immediately obvious.  Such as how we place certain things under the built-in section rather than the language section, when it's not immediately obvious where they might be found.

File: desktop-src/direct3dhlsl_current/hlsl-language-access_qualifiers.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# Access Qualifiers
+
+See [`DirectX Spec`](https://microsoft.github.io/DirectX-Specs/) for more details.
```
Comment (tex3d) [inline desktop-src/direct3dhlsl_current/hlsl-language-access_qualifiers.md:3]: What's this page for?  We already have: `hlsl-built-ins-raytracing-payload-access-qualifiers.md`.  This just links to the top-level spec repo, so it's very unhelpful.  This page is probably lefover and should be deleted.

File: desktop-src/direct3dhlsl_current/hlsl-language-attributes.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+# Attributes
+
+Attributes describe additional behavior for language features.  These attributes precede language features and are enclosed in square brackets.
```
Comment (tex3d) [inline desktop-src/direct3dhlsl_current/hlsl-language-attributes.md:3]: I don't think this is nearly sufficient for attributes.  Ideally, we should break down the different attribute types that HLSL supports, show the syntax (that's the language part), and link to the built-in section for the set of attributes recognized in each location.

For instance:
- Function Attributes: before a function declaration/definition
- Loop Attributes: before a for/do/while statement
- Branch Attributes: before an if statement
- Switch Attributes: before a switch statement
- etc...

File: desktop-src/direct3dhlsl_current/hlsl-built-ins-functions_math_logic.md [hunk 4]
```diff
@@ -0,0 +1,3332 @@
+# Math and Logic Operations
+
+### abs
```
Comment (tex3d) [inline desktop-src/direct3dhlsl_current/hlsl-built-ins-functions_math_logic.md:3]: If all these are going to be described on the same page, we at least need TOC nodes for navigation.  It's also visually hard to parse and see where one operation ends and another begins.  I wish the content under an operation was all indented or something, so it would be visually obvious.

Comment (tex3d) [reply desktop-src/direct3dhlsl_current/hlsl-built-ins-functions_math_logic.md:3]: Could parameter, return value and type tables be merged into one table by simply adding the description to the type table?

I imagine this would need to be done with a script, otherwise it's a lot of mechanical labor.

File: desktop-src/direct3dhlsl_current/hlsl-language-functions.md [hunk 5]
```diff
@@ -0,0 +1,350 @@
+# Functions
+
+Functions encapsulate HLSL statements. This enables you to debug a set of functions and then reuse them across shaders or effects. You may want to create a function that encapsulates the functionality of a vertex shader, pixel shader or texture shader. Other times, you may want to write a helper function that performs some commonly used task, and then call that helper function from your shader function. The rules for writing shader functions for HLSL are very similar to writing C functions.
+
+## Function Declaration Syntax
+
+```syntax
+[StorageClass] [clipplanes()] [precise] Return_Value Name ( [ArgumentList] ) [: Semantic] {   [StatementBlock] };
```
Comment (tex3d) [inline desktop-src/direct3dhlsl_current/hlsl-language-functions.md:8]: `[clipplanes()]` should be replaced with `[FunctionAttributes]`.

File: desktop-src/direct3dhlsl_current/hlsl-language-functions.md [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+```
+
+### Parameters
+
+#### `StorageClass`
+
+Modifier that redefines a function declaration. `inline` is currently the only modifier value. The modifier value must be `inline` because it is also the default value. Therefore, a function is inline regardless of whether you specify `inline`, and all functions in HLSL are inline. An inline function generates a copy of the function body (when compiling) for each function call. This is done to decrease the overhead of calling the function.
+
+#### `Clipplanes`
```
Comment (tex3d) [inline desktop-src/direct3dhlsl_current/hlsl-language-functions.md:17]: This should be replaced with `FunctionAttributes`.  `[clipplanes()]` is just one function attribute valid on the vertex shader entry point specifically.  There are many more function attributes, which should be described under the built-ins section for the relevant contexts.

General attribute syntax should be shown here, since it's not clear from the above what these would look like.

We should link to a `FunctionAttributes` section on a general `HLSLAttributes` page describing the various kinds of attributes, linking to the built-in attributes for each kind.

The general syntax for HLSL attributes is something like:
```
HLSLAttributes : HLSLAttribute [ "," HLSLAttributes ]
HLSLAttribute : "[" Name [ "(" Arguments ")" ] "]"
Name : Identifier
Arguments : Argument [ "," Arguments ]
Argument : depends on the attribute
```
Though we don't use this sort of grammar convention for syntax in the docs - I'm just not sure how explicit `[` and `]` can be clearly shown with the syntax for function above.

File: desktop-src/direct3dhlsl_current/hlsl-method-getDimensions.md [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+
+None.
+
+## Example
+
+```HLSL
+// None.
+```
+
+## Supported Shader Models
```
Comment (tex3d) [inline desktop-src/direct3dhlsl_current/hlsl-method-getDimensions.md:329]: This table looks about right for Sample() and some related methods, but not for GetDimensions().  GetDimensions() should be supported at every stage on every shader model.

File: desktop-src/direct3dhlsl_current/hlsl-method-sample.md [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+| Mesh | - | - | - | - | - | - | 1* |
+| Pixel | x | x | x | x | x | x | x |
+| Compute | - | - | - | - | - | - | x |
+
+| Key | Description |
+| - | - |
+| x | Supported |
+| 1* | Supported with optional feature: `SHADER_FEATURE_DERIVATIVES_IN_MESH_AND_AMPLIFICATION_SHADERS` |
+
+>TBD: Make optional feature into a link
```
Comment (tex3d) [inline desktop-src/direct3dhlsl_current/hlsl-method-sample.md:224]: This TBD should be resolved and removed from here.  It's replicated all over the place, along with the table that isn't correct for most other operations.

File: desktop-src/direct3dhlsl_current/hlsl-method-sample.md [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+   [, out uint          status ] ] ] );
+```
+
+#### TextureCube
+```syntax
+ElementType TextureCube<ElementType>::Sample(
+      in  SamplerState  samp,
+      in  float3        coord
+   [, in  float         lodClamp
+   [, out uint          status ] ] ] );
```
Comment (tex3d) [inline desktop-src/direct3dhlsl_current/hlsl-method-sample.md:66]: typo: extra square close bracket `]`.  Same with next method signature.

File: desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+   [, out uint          status ] );
+
+ElementType Texture2D<ElementType>::GatherCmpRed(
+      in  SamplerState  samp,
+      in  float2        coord,
+      in  float         compareValue,
+      in  int2          offset,
+   [, in  int2          offset2]
+   [, in  int2          offset3]
+   [, in  int2          offset4]
```
Comment (tex3d) [inline desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:26]: Offset2/3/4 aren't independently optional.  You can either use one offset, or 4 offsets.  So perhaps the syntax here should look more like this:
```
   [, in  int2          offset2
    , in  int2          offset3
    , in  int2          offset4]
```

Comment (tex3d) [reply desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:26]: Oh and same for all similar places here and for normal Gathers

File: desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+      in  SamplerState  samp,
+      in  float2        coord,
+      in  float         compareValue,
+      in  int2          offset,
+   [, in  int2          offset2]
+   [, in  int2          offset3]
+   [, in  int2          offset4]   
+   [, out uint          status ] );
+
+ElementType Texture2D<ElementType>::GatherCmpGreen(
```
Comment (tex3d) [inline desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:29]: Do we need to expand out each signature for Red, Green, Blue, and Alpha?  This is just component selection on identical method overloads.  Could the method name be written: `GatherCmp[Red|Green|Blue|Alpha]` with one signature instead?  (same for normal Gather)

File: desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+This provides the HLSL type for an element loaded from a texture.
+The format stored in the resource is based on the texture resource view's DXGI_FORMAT,
+which will determine how data is translated into the type specified in HLSL.
+
+## Parameters
+
+
+| Parameter | Description |
+| - | - |
+| in [`SamplerState samp`](#samplerstate-samp) | SamplerState object that determines how sampling will take place. |
```
Comment (tex3d) [inline desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:191]: This should be `SamplerComparisonState` for `GatherCmp*`.

File: desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+| TextureObject | [`<CoordType>`](#coordtype-coord) | [`<OffsetType>`](#offsettype-offset) |
+| --- | --- | --- |
+| `Texture2D` | `float2` | `int2` |
+| `Texture2DArray` | `float3` | `int2` |
+| `TextureCube` | `float3` | N/A |
+| `TextureCubeArray` | `float4` | N/A |
+
+### `SamplerState samp`
+
+A Sampler state. This is an object declared in an effect file that contains state assignments.
```
Comment (tex3d) [inline desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:208]: SamplerComparisonState...

File: desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+
+> My alternate text:
+> SamplerState object that determines how sampling will take place.
```
Comment (tex3d) [inline desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:211]: Should have resolved these and "TBD" items instead of replicating them elsewhere.

File: desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+| `Texture2D` | `int2` | xy |
+| `Texture2DArray` | `int2` | xy |
+
+### `uint status`
+
+The status of the operation. You can't access the status directly; instead, pass the status to the CheckAccessFullyMapped intrinsic function. CheckAccessFullyMapped returns TRUE if all values from the corresponding Sample, Gather, or Load operation accessed mapped tiles in a tiled resource. If any values were taken from an unmapped tile, CheckAccessFullyMapped returns FALSE.
+
+## Remarks
+
+None.
```
Comment (tex3d) [inline desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:254]: Nothing describes what Red/Green/Blue/Alpha means in the method name.  Wouldn't this be a suitable place for that?

File: desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+
+
+## Example
+
+```HLSL
+// N/A
```
Comment (tex3d) [inline desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:260]: Should probably have an example if we have a section for it.

File: desktop-src/direct3dhlsl_current/hlsl-built-ins-functions_math_logic.md [hunk 17]
```diff
@@ -0,0 +1,3332 @@
...
+
+#### Return Value
+
+The absolute value of the *x* parameter.
+
+#### Type Description
+
+
+
+| Name  | [**Template Type**](dx-graphics-hlsl-intrinsic-functions.md)                                                  | [**Component Type**](dx-graphics-hlsl-intrinsic-functions.md)                 | Size                           |
```
Comment (tex3d) [inline desktop-src/direct3dhlsl_current/hlsl-built-ins-functions_math_logic.md:36]: These links to `dx-graphics-hlsl-intrinsic-functions.md` aren't right.  Looks like this was replicated throughout this file.

File: desktop-src/direct3dhlsl_current/hlsl-built-ins-functions_math_logic.md [hunk 18]
```diff
@@ -0,0 +1,3332 @@
...
+|-------|----------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------|--------------------------------|
+| *x*   | [**scalar**](dx-graphics-hlsl-intrinsic-functions.md), **vector**, or **matrix** | [**float**](/windows/desktop/WinProg/windows-data-types), [**int**](/windows/desktop/WinProg/windows-data-types) | any                            |
```
Comment (tex3d) [inline desktop-src/direct3dhlsl_current/hlsl-built-ins-functions_math_logic.md:38]: If types are linked, they should probably link to the appropriate HLSL type definitions, not windows-data-types, which doesn't even define these types, and instead defines a bunch of types from Windows headers that are not applicable to HLSL at all.  For now, I'd say just get rid of the links here.

Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-built-ins-functions_math_logic.md:38]: Alright, fixed links to the proper locations in the current documentation.

File: desktop-src/direct3dhlsl_current/hlsl-built-ins-functions_math_logic.md (hunk 17 above)
Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-built-ins-functions_math_logic.md:36]: Fixed.

File: desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md (hunk 16 above)
Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:260]: I agree.  If you can find an example for this method, please let me know.  For completeness, every section has an example even if it is blank.

File: desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md (hunk 15 above)
Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:254]: I have added explanation for gather.

File: desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md (hunk 14 above)
Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:211]: Great alternate text here.

Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:211]: Replaced all instances with links added.

File: desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md (hunk 13 above)
Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:208]: Done. Thanks.

File: desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md (hunk 12 above)
Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:191]: Thank you. Done.

File: desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md (hunk 11 above)
Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:29]: I see what you are saying.  It seems like a good place to spell out everything is in the function signatures.  Of course, can easily correct if needed.

File: desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md (hunk 10 above)
Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:26]: I will also include in the introduction that the signatures are valid only with the first offset value or all 4.  There are no other signatures.

Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:26]: Added in your alternate text.

File: desktop-src/direct3dhlsl_current/hlsl-method-sample.md (hunk 9 above)
Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-method-sample.md:66]: Thank you.  Corrected.

File: desktop-src/direct3dhlsl_current/hlsl-method-sample.md (hunk 8 above)
Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-method-sample.md:224]: I think we should remove these compatibility tables until we have a good method of researching compatibility.

File: desktop-src/direct3dhlsl_current/hlsl-method-getDimensions.md (hunk 7 above)
Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-method-getDimensions.md:329]: We should remove compatibility currently.

File: desktop-src/direct3dhlsl_current/hlsl-language-functions.md (hunk 6 above)
Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-language-functions.md:17]: There is some sort of error in the built-in attributes section.  I will put in clip planes and debug in the process.

File: desktop-src/direct3dhlsl_current/hlsl-language-functions.md (hunk 5 above)
Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-language-functions.md:8]: also added SV_ClipDistance to System Value semantics.

File: desktop-src/direct3dhlsl_current/hlsl-language-functions.md (hunk 6 above)
Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-language-functions.md:17]: Added attributes section linking to both the language definition as well as the built-in attributes section.

File: desktop-src/direct3dhlsl_current/hlsl-built-ins-functions_math_logic.md (hunk 4 above)
Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-built-ins-functions_math_logic.md:3]: I could write a simple C# program with regular expressions.  It won't be too complicated.

File: desktop-src/direct3dhlsl_current/hlsl-language-access_qualifiers.md (hunk 2 above)
Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-language-access_qualifiers.md:3]: For the time being, there are only raytracing payload access qualifiers, so I removed this section from the repo.

File: desktop-src/direct3dhlsl_current/hlsl-landing-page.md (hunk 1 above)
Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-landing-page.md:1]: Done.

File: desktop-src/direct3dhlsl_current/hlsl-language-functions.md (hunk 5 above)
Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-language-functions.md:8]: Done.  Put clipplanes in the built-in attributes and changed the clipplanes into attributes in the language section.

File: desktop-src/direct3dhlsl_current/hlsl-language-attributes.md (hunk 3 above)
Comment (toddmacedo) [reply desktop-src/direct3dhlsl_current/hlsl-language-attributes.md:3]: Alright, wrote original copy and headers for these sections.  I expanded flow control into loop attributes and branch attributes.

**Comment by toddmacedo** [comment]
Links Fixed:
•	HLSL Language
o	Flow Control
	Do statement
	For statement
	While statement
o	Functions
	Code sample link does not work in original as well.
	BasicHLSL10.fx
	AdvancedParticles.fx
	BasicHLSL10Sample
	ContentStreaming Sample
o	Statements
	Expressions
•	Variables goes to variable syntax
•	Added operator link
	Added additional
•	Variables
•	Operators
	Added math and logic links
•	All
•	Any
•	Asfloat
•	Asint
•	Asdouble
	Binary Casts
•	Asfloat
•	Asint
•	Asuint
	Bitwise operators
•	Fixed link to sm 4.0
	Comparison operators
•	Any
•	All
	Statement Blocks
o	User-defined types
	Enumerations
•	Code block at end
	Structs
•	HLSL packing rules redirect
•	Shader model 4 redirect
o	Attributes
	Entry function attributes link
	Branching attributes link
o	Packoffset
	Link to declaring a variable type
o	Register
	Link to shader profile
	Packoffset
o	Global Blocks
	Add link to register
	Add link to packoffset
	Added variable links
	Added vector links
	Added arrays
	Fixed DX11 links
o	Root signatures
	D3D_ROOT_SIGNATURE_FLAGS fixed
•	Built-ins
o	Preprocessor directives
	#define link
	Pragma directives
•	Missing 
o	Def
o	Message
o	Pack_matrix
o	Constants
	Added shader section.
o	Math and logic functions
	](intrinsic functions)
o	Raytracing functions
	Intersection Attributes structure
o	Attributes
	Link in intro
	SV_GSInstanceID
	ID3D11DeviceContext::Dispatch
o	Semantics
	Multiple links fixed.
•	Appendix
o	HLSL Language Versions
	Enum and enum class fixed.
o	Shader Model page
	Fixed links in high-level page
